README.md
Last Updated: 12/10/2013 07:00pm  
(see CHANGELOG for details)

### RFduino: Shrunk down an Arduino to the size of a finger-tip and made it Wireless!

![logo](https://raw.github.com/RFduino/RFduino/master/4up%20image.jpg)  

[Video](http://www.youtube.com/watch?v=arWBdGwCJcM)  

### Installation

* Get the hardware here: [http://RFduino.com](http://RFduino.com)

* Download Arduino 1.5 here: [http://arduino.cc](http://arduino.cc/en/Main/Software)  
  (on osx remember to open arduino first to make gatekeeper perform its magic)  

* Copy the RFduino directory from this repository in Arduino  
  (on Windows, C:\arduino-1.5.4\hardware\arduino)  
  (on OSX, /Applications/Arduino.app/Contents/Resources/Java/hardware/arduino)  
  or "git clone https://github.com/RFduino/RFduino" in the directory indicated

* Install the FTDI drivers from here: [http://ftdichip.com](http://www.ftdichip.com/Drivers/VCP.htm)

Your ready to go!

Detailed instructions are available here: [Quick Start Guide](http://files.rfdigital.com/rfduino.quick.start.guide.pdf)

### Getting Started

* Attach the USB shield

* Select RFduino from the Tools/Board menu

* Select the port from the Tools/Port menu

* Select an example from the Files/Examples/RFduinoNonBLE or Files/Examples/RFduinoBLE directory

* Select Upload to compile, upload and execute the sketch

* Down the iPhone example apps from the iPhone App Store (search for "RFduino").

Detailed instructions for the Temperature app are available here: [Temperature App](http://files.rfdigital.com/rfduino.temperature.guide.pdf)  
Detailed instructions for the ColorWheel app are available here: [ColorWheel App](http://files.rfdigital.com/rfduino.rgb.colorwheel.guide.pdf)  

### Communicating with us

The best way to communiate with us is here: [RFduino Forum](http://forum.RFduino.com).

### A Lot More Coming!

We are working on documentation as fast as we can.  Until then, the best source for documentation is the sketch examples.

This project has been a huge success, and we have many exciting things planned that we want to share with the community.  At the same time, we are unbelievable busy, with a ton going on and lots of small items to clean up.  We are just a small team working on this project ... please bear with us!

We hope you enjoy creating stuff with your RFduino as much as we do!

The RFduino team.



CHANGELOG
v 1.4
  - Integrated RFDLoader

v 1.3.2
  - Fixed issue with the send characteristic always sending 20 bytes
  - Increased send characteristic payload size from 12 to the maximum of 20 bytes
  - Added sketch and iPhone demonstrating lossless bulk data transfer

v 1.3.1
  - Fixed issue with on_rssi() only working for the first connect

v 1.3.0
  - Added support for radio active notification to support sketches with critical timing requirements.

v 1.2.0
  - Restructured the RFduino repository to allow a git clone directly into the Arduino directory structure.

v 1.1.2
  - Corrected the name of a missing extern.
  - UART_DEFAULT_RX_PIN and UART_DEFAULT_TX_PIN and multiply defined (and conflicting).  Removed the erroneous definitions.

v 1.1.1
  - Added support for OSX Snow Leopard.

v 1.1.0
  - Support for OSX.
  - equivalent to RFduino305.zip.

v 1.0.0
  - equivalent to RFduino289.zip.


RFduino/cores/arduino/UARTClass.h
/*
 Copyright (c) 2013 OpenSourceRF.com.  All right reserved.

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/*
  Copyright (c) 2011 Arduino.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef _UART_CLASS_
#define _UART_CLASS_

#include "HardwareSerial.h"
#include "RingBuffer.h"
//#include "variant.h"

// Includes Atmel CMSIS
#include <chip.h>

#define UART_DEFAULT_RX_PIN		1
#define UART_DEFAULT_TX_PIN		0

typedef void Uart;

class UARTClass : public HardwareSerial
{
  protected:
    RingBuffer *_rx_buffer ;

  public:
    UARTClass( RingBuffer* pRx_buffer ) ;

    void begin( const uint32_t dwBaudRate ) ;
	void begin( const uint32_t dwBaudRate, uint8_t rx_pin, uint8_t tx_pin ) ;
    void end( void ) ;
    int available( void ) ;
    int peek( void ) ;
    int read( void ) ;
    void flush( void ) ;
    size_t write( const uint8_t c ) ;

    void IrqHandler( void ) ;

#if defined __GNUC__ /* GCC CS3 */
    using Print::write ; // pull in write(str) and write(buf, size) from Print
#elif defined __ICCARM__ /* IAR Ewarm 5.41+ */
//    virtual void write( const char *str ) ;
//    virtual void write( const uint8_t *buffer, size_t size ) ;
#endif

    operator bool() { return true; }; // UART always active
};

#endif // _UART_CLASS_


libraries/RFduinoBLE/RFduinoBLE.cpp
/*
 Copyright (c) 2013 OpenSourceRF.com.  All right reserved.

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

#include "Arduino.h"
#include "RFduinoBLE.h"

RFduinoBLEClass::RFduinoBLEClass()
{
	deviceName = "RFduino";
	advertisementData = "sketch";
	advertisementInterval = 80;
	txPowerLevel = +4;
}

int RFduinoBLEClass::begin()
{
  RFduinoBLE_device_name = deviceName;
  RFduinoBLE_advertisement_data = advertisementData;
  RFduinoBLE_advertisement_interval = advertisementInterval;
  RFduinoBLE_tx_power_level = txPowerLevel;

  return RFduinoBLE_begin();
}

void RFduinoBLEClass::end()
{
  RFduinoBLE_end();
}

bool RFduinoBLEClass::send(char data)
{
	return RFduinoBLE_send(&data, 1);
}

bool RFduinoBLEClass::send(const char *data, int len)
{
	return RFduinoBLE_send(data, len);
}

bool RFduinoBLEClass::sendByte(uint8_t data)
{
  return RFduinoBLE_send((char*)&data, 1);
}

bool RFduinoBLEClass::sendInt(int data)
{
  return RFduinoBLE_send((char*)&data, sizeof(int));
}

bool RFduinoBLEClass::sendFloat(float data)
{
  return RFduinoBLE_send((char*)&data, sizeof(float));
}

RFduinoBLEClass RFduinoBLE;


RFduino/variants/RFduino/libRFduino.h
#ifndef _LIBRFDUINO_H_
#define _LIBRFDUINO_H_

#ifdef __cplusplus
extern "C" {
#endif // __cplusplus

#define MILLISECONDS(x) ((uint64_t)x)
#define SECONDS(x)      (x * MILLISECONDS(1000))
#define MINUTES(x)      (x * SECONDS(60))
#define HOURS(x)        (x * MINUTES(60))
#define DAYS(x)         (x * HOURS(24))
#define INFINITE (-1)

#define CELSIUS 0
#define FAHRENHEIT 1

#define LOW 0
#define HIGH 1
#define DISABLE 5

// pass the pin number so you can use the same interrupt handler for all pins
// return 0 to reset wake_detect bit (and not exit RFduino_ULPDelay)
// return 1 to leave wake_detect bit alone (and exit RFduino_ULPDelay)
typedef int (*pin_callback_t)(uint32_t ulPin);

/**
 * \brief Ultra low power delay for the amount of time (in milliseconds) specified as parameter.
 *        You can use the following macros: MILLISECONDS(x), SECONDS(x), MINUTES(x), HOURS(x), DAY(x), INFINITE
 *        ie: RFduino_ULPDelay( MINUTES(2) + SECONDS(30) )
 *
 *        RFduino_pinWake() can be used to cause HIGH / LOW edges on GPIO pins to exit ulp_delay.
 *        see documentation for RFduino_pinWake for more details.
 *
 * \param ms the number of milliseconds to switch to ultra low power (max is DAYS(7) unless INFINITE)
 */
extern void RFduino_ULPDelay( uint64_t ms );

// the same function is used for both BLE and NonBLE (so alias it both ways)
#define RFduinoBLE_ULPDelay(x) RFduino_ULPDelay(x)

/**
 * \brief Configures the pin to wake the device from system off.
 *
 * \param ulPin The number of the digital pin you want to read (int)
 * \param dwWake Either DISABLED, HIGH or LOW
 */
extern void RFduino_pinWake( uint32_t ulPin, uint32_t dwWake );

/**
 * \brief Test whether a pin caused us to wakeup and exit ulp_delay
 *
 * \param ulPin The number of the digital pin you want to read (int)
 *
 *        You must call RFduino_resetPinWoke() to reset this condition after processing
 */
extern int RFduino_pinWoke( uint32_t ulPin );

/**
 * \brief Reset state of pin that caused us to wakeup and exit ulp_delay
 *
 * \param ulPin The number of the digital pin you want to reset
 */
extern void RFduino_resetPinWake( uint32_t ulPin );

/**
 * \brief Configures the pin to wake the device from system off and execute a callback.
 *
 * \param ulPin The number of the digital pin you want to read (int)
 * \param dwWake Either DISABLED, HIGH or LOW
 * \param callback The callback function to execute when the pin is detected.
 *                 Here is an example pin callback function:
 *                 int my_pin_callback(uint32_t pin)
 *                 {
 *                   // do something
 *                   // return 0 to reset wake_detect, return 1 to exit RFduino_ULPDelay
 *                   return 0;
 *                 }
 *                 To call this function on the when pin 5 goes high:
 *                 RFduino_pinWake(5, HIGH, my_pin_callback);
 */
extern void RFduino_pinWakeCallback( uint32_t ulPin, uint32_t dwWake, pin_callback_t callback ) ;

/**
 * \brief Reset the system
 */
extern void RFduino_systemReset(void);

/**
 * \brief Turn the system off
 */
extern void RFduino_systemOff(void);

/**
 * \brief Get a sample from the temperature sensor
 *
 * \param scale Either CELSIUS or FAHRENHEIT
 *
 * returns:
 *     degrees c (-198.00 to +260.00)
 *     degrees f (-128.00 to +127.00)
 */
float RFduino_temperature(int scale);

uint32_t rfduino_ppi_channel_assign(uint8_t channel_num, const volatile void * event_endpoint, const volatile void * task_endpoint);
uint32_t rfduino_ppi_channel_unassign(uint8_t channel_num);

uint32_t rfduino_nvic_enableirq(uint8_t irq_num);
uint32_t rfduino_nvic_disableirq(uint8_t irq_num);

#ifdef __cplusplus
}
#endif

#endif


RFduino/libraries/RFduinoBLE/RFduinoBLE.h
/*
 Copyright (c) 2013 OpenSourceRF.com.  All right reserved.

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

#ifndef _RFDUINOBLE_H_
#define _RFDUINOBLE_H_

#include <stdint.h>
#include <stdbool.h>

// needed for callback support
#include "libRFduinoBLE.h"

class RFduinoBLEClass
{
public:
  // default "RFduino"
  const char *deviceName;

  // default "sketch"
  const char *advertisementData;

  // in ms (converted to 0.625ms units) - range 20ms to 10.24s - default 80ms
  int advertisementInterval;

  // -20 dBm to +4 dBm - default +4 dBm
  int txPowerLevel;

public:
  RFduinoBLEClass();

  int begin();
  void end();

  bool send(char data);
  bool send(const char *data, int len);

  bool sendByte(uint8_t data);
  bool sendInt(int data);
  bool sendFloat(float data);
};

extern RFduinoBLEClass RFduinoBLE;

#endif


variants/RFduino/variant.h
/*
 Copyright (c) 2013 OpenSourceRF.com.  All right reserved.

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/*
  Copyright (c) 2011 Arduino.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

/*
 * Here should go some sort of liscense maybe? This file is a modified
 * arduino source file.  It defines the board variant being used
*/

#ifndef _VARIANT_RFDUINO_
#define _VARIANT_RFDUINO_

/*----------------------------------------------------------------------------
 *        Headers
 *----------------------------------------------------------------------------*/

#include "Arduino.h"

#ifdef __cplusplus
#include "UARTClass.h"
#endif

#ifdef __cplusplus
extern "C"{
#endif // __cplusplus

/**
 * Libc porting layers
 */
#if defined (  __GNUC__  ) /* GCC CS3 */
#    include <syscalls.h> /** RedHat Newlib minimal stub */
#endif

/*----------------------------------------------------------------------------
 *        Definitions
 *----------------------------------------------------------------------------*/

/*----------------------------------------------------------------------------*/

#define RFDUINO

/** Name of the board */
#define VARIANT_NAME        "RFduino"

/** Master clock frequency  (used in TWI and DAC) */
#define VARIANT_MCK         16000000UL

/*----------------------------------------------------------------------------
 *        Pins
 *----------------------------------------------------------------------------*/

// Number of pins
#define PINS_COUNT           (7u)

/* Pins descriptions to be instantiated into variant.cpp */
extern uint8_t PPI_Channels_Occupied[PINS_COUNT][2];
extern uint8_t GPIOTE_Channels_Occupied[PINS_COUNT];
extern uint8_t Timer1_Compare_Unit_Occupied_by_Pin[3]; // Contains Pin number which occupied appropriate CC register (0, 1 or 2) of Timer 1 for PWM purposes
extern uint8_t Timer2_Compare_Unit_Occupied_by_Pin[3]; // Contains Pin number which occupied appropriate CC register (0, 1 or 2) of Timer 2 for PWM purposes
extern uint8_t Pin_Occupied_for_PWM[PINS_COUNT]; // Determines if particular pin is occupied for PWM output

/*
 * SPI Interfaces
 */
#define SPI_INTERFACES_COUNT 2

#define SPI_INTERFACE        NRF_SPI0
#define PIN_SPI_SS           (6u)
#define PIN_SPI_MOSI         (5u)
#define PIN_SPI_MISO         (3u)
#define PIN_SPI_SCK          (4u)

#define SS    PIN_SPI_SS
#define MOSI  PIN_SPI_MOSI
#define MISO  PIN_SPI_MISO
#define SCK   PIN_SPI_SCK

/*
 * Wire Interfaces
 */
#define PIN_WIRE_SDA         (6u)
#define PIN_WIRE_SCL         (5u)
#define WIRE_INTERFACE       NRF_TWI1

/*
 * Analog pins
 */
#define ADC_RESOLUTION        10

/*
 * UART/USART Interfaces
 */
#define UART_DEFAULT_RX_PIN		0
#define UART_DEFAULT_TX_PIN		1

void UART0_Start( int dwBaudRate, uint8_t rx_pin, uint8_t tx_pin );
void UART0_Stop();
void UART0_FlushTX();
void UART0_TX( const uint8_t uc_data );

__attribute__(( always_inline ))
static int __INLINE UART0_RXReady()
{
  return NRF_UART0->EVENTS_RXDRDY;
}

__attribute__(( always_inline ))
static uint8_t __INLINE UART0_RXData()
{
  return NRF_UART0->RXD;
}

void UART0_RXReset( int isr );
uint8_t UART0_RX();

/*
 * PWM
 */
#define PWM_MAX_DUTY_CYCLE    255
#define PWM_MIN_DUTY_CYCLE    0
#define PWM_RESOLUTION        16

/*
 * Miscellaneous
 */

#ifdef __cplusplus
}
#endif

/*----------------------------------------------------------------------------
 *        Arduino objects - C++ only
 *----------------------------------------------------------------------------*/

#ifdef __cplusplus

extern UARTClass Serial;

#endif

#endif /* _VARIANT_RFDUINO_ */
